Arm64 Kernel Module Updates & New Config Fields

cwshugg opened the pull request:
This PR implements a number of new config fields to allow for the soft-coding of various external executables used by Revizor (such as `as`, `gcc`, `objdump`, `awk`, `modprobe`, etc.). Different OSs that Revizor runs on may have different names/paths for common executables; hard-coding these executable names can be an obstacle for getting the fuzzer running on new systems. The default values for these new config fields should reflect the existing hard-coded values, except for a few Arm64-specific executables (whose proper values have been added to the Arm64 test YAMLs).

This PR also adds preprocessor guards to the Arm64 executor kernel module, to allow for the choice between the Flush+Reload and Prime+Probe attack. Currently, since Prime+Probe isn't fully functioning, its preprocessor flag is disabled (meaning the kernel module by default will use the Flush+Reload attack).

OleksiiOleksenko commented on src/arm64/executor/main.c:
```diff
@@ -42,9 +42,14 @@ int (*set_memory_nx)(unsigned long, int) = 0;
-char *measurement_template = (char *)&template_l1d_flush_reload; // template_l1d_prime_probe is an alternative but is currently non-functional
 char *measurement_code = NULL;
 
+#if defined(CACHE_TIMING_ATTACK_FLUSH_RELOAD)
```
The module already has a function for selecting the attack type (see `measurement_mode_store`), so the preprocessor macro duplicates this functionality. You can find a reference implementation in the `x86` directory

cwshugg replied:
Ah, thank you for pointing this out - I will update `measurement_mode_store` to look for specific characters in the buffer to select an attack type, like I see in the x86 executor.

OleksiiOleksenko commented on src/arm64/tests/test-detection.yaml:
```diff
@@ -8,8 +8,6 @@ min_bb_per_function: 2
 avg_mem_accesses: 4
 input_gen_entropy_bits: 9
-
-enable_ssbp_patch: true
```
Doesn't ARM have a patch against Speculative Store Bypass (SSBP)?

cwshugg replied:
Yes, Arm has the Physical SSBP Barrier instruction ([PSSBB](https://developer.arm.com/documentation/ddi0597/2021-03/Base-Instructions/PSSBB--Physical-Speculative-Store-Bypass-Barrier-?lang=en)) and the "Speculative Store Bypass Safe" register ([SSBS](https://developer.arm.com/documentation/ddi0601/2023-09/AArch64-Registers/SSBS--Speculative-Store-Bypass-Safe)), which toggles their patch.

I removed this field from the YAML because I couldn't find it being used anywhere in the Arm port code. (Last year, we didn't implement it as part of the Hackathon. Unless it was added since then, support for that does not exist in the Arm port.) My intention was to clean up the test YAML, but I'm happy to add this back in if you'd prefer it be left alone until Arm SSBP support can be added in.

OleksiiOleksenko commented:
@cwshugg I've left some comments in the code review. Could you take a look?

cwshugg commented:
@OleksiiOleksenko thanks for taking a look at this; I removed the preprocessor defines and replied to your comment about the Arm SSBP patch. Let me know what you think!

OleksiiOleksenko commented on src/arm64/executor/main.c:
```diff
@@ -42,9 +42,11 @@ int (*set_memory_nx)(unsigned long, int) = 0;
 char *measurement_code = NULL;
 
+char *measurement_template = (char *)&template_l1d_flush_reload;
+char *measurement_template = (char *)&template_l1d_prime_probe;
```
This is a duplicate declaration of a global variable. I'd guess, this won't compile

cwshugg replied:
Ah! Thank you. A quick delete and failure to check the resulting code on my part. The duplicate has been removed.

OleksiiOleksenko commented on src/arm64/executor/main.c:
```diff
@@ -361,9 +363,17 @@ static ssize_t enable_pre_run_flush_store(struct kobject *kobj, struct kobj_attr
+    else
     {
-        measurement_template = (char *)&template_l1d_flush_reload;  // template_l1d_prime_probe is an alternative but is currently non-functional
+        printk(KERN_ERR "arm64_executor: Neither l1d_flush_reload nor l1d_prime_probe is enabled.\n");
```
Return error code (-1) to make the error more explicit

cwshugg replied:
Good idea. Return code added!

OleksiiOleksenko commented:
> @OleksiiOleksenko thanks for taking a look at this; I removed the preprocessor defines and replied to your comment about the Arm SSBP patch. Let me know what you think!

Thanks! Looks good, just two minor issues left (see my comments)

cwshugg commented:
> > @OleksiiOleksenko thanks for taking a look at this; I removed the preprocessor defines and replied to your comment about the Arm SSBP patch. Let me know what you think!
> 
> @cwshugg Thanks! Looks good, just two minor issues left (see my comments)

@OleksiiOleksenko thank you! I fixed up the two issues you found.

OleksiiOleksenko reviewed (approved):
All good, ready to merge
